### PR TITLE
Provision Resources to Deploy EPiServerCMS in Azure

### PR DESCRIPTION
**Pull request by rajbalakrish**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@rajbalakrish__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

**Comment by rajbalakrish** [comment]
Pl. review PR and feedback if everything is OK

**Comment by azurecla** [comment]
__@rajbalakrish__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

**Comment by rajbalakrish** [comment]
@gatneil Could you review this Pull Request please

**Comment by gatneil** [comment]
Please rename the folder to 'episerver'. A full solution like this shouldn't be numbered, and it should be clear from context that it's on Azure :)

**Comment by gatneil** [comment]
There are a lot of parameters with no default values; please either give them default values or make them variables. In the case of storage account names, please see the best practice around auto-generating storage account names: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

File: 101-episerver-in-azure/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+      "storageId":"[resourceId('Microsoft.Storage/storageAccounts', parameters('StorageAccountName'))]",
+      "storageConnectionStringPrefix":"[concat('DefaultEndpointsProtocol=https;AccountName=',parameters('StorageAccountName'),';AccountKey=')]"
+   },
+   "resources":[
+      {
+         "name":"[parameters('WebApplication-HostingPlanNameName')]",
+         "type":"Microsoft.Web/serverfarms",
+         "location":"[resourceGroup().location]",
+         "apiVersion":"2014-06-01",
+         "dependsOn":[
```
Comment (gatneil) [inline 101-episerver-in-azure/azuredeploy.json]: If there are no dependsOn, can we remove this empty list?

File: 101-episerver-in-azure/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+         "properties":{
+            "accountType":"[parameters('StorageAccountType')]"
+         }
+      },
+      {
+         "name":"[parameters('sqlserverName')]",
+         "type":"Microsoft.Sql/servers",
+         "location":"[resourceGroup().location]",
+         "apiVersion":"2014-04-01-preview",
+         "dependsOn":[
```
Comment (gatneil) [inline 101-episerver-in-azure/azuredeploy.json]: same

File: 101-episerver-in-azure/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+         },
+         "properties":{
+            "accountType":"[parameters('StorageAccountType')]"
+         }
+      },
+      {
+         "name":"[parameters('sqlserverName')]",
+         "type":"Microsoft.Sql/servers",
+         "location":"[resourceGroup().location]",
+         "apiVersion":"2014-04-01-preview",
```
Comment (gatneil) [inline 101-episerver-in-azure/azuredeploy.json]: is there a non-preview api version we can use?

File: 101-episerver-in-azure/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+            "administratorLogin":"[parameters('sqlserverAdminLogin')]",
+            "administratorLoginPassword":"[parameters('sqlserverAdminLoginPassword')]",
+            "version":"12.0"
+         },
+         "resources":[
+            {
+               "name":"AllowAllWindowsAzureIps",
+               "type":"firewallrules",
+               "location":"[resourceGroup().location]",
+               "apiVersion":"2014-04-01-preview",
```
Comment (gatneil) [inline 101-episerver-in-azure/azuredeploy.json]: same

File: 101-episerver-in-azure/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+               "properties":{
+                  "startIpAddress":"0.0.0.0",
+                  "endIpAddress":"0.0.0.0"
+               }
+            },
+            {
+               "name":"[parameters('SQL-DatabaseName')]",
+               "type":"databases",
+               "location":"[resourceGroup().location]",
+               "apiVersion":"2014-04-01-preview",
```
Comment (gatneil) [inline 101-episerver-in-azure/azuredeploy.json]: same

File: 101-episerver-in-azure/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+   "outputs":{
+      "NamespaceDefaultConnectionString":{
+         "type":"string",
+         "value":"[listkeys(variables('defaultAuthRuleResourceId'), variables('serviceBusApiVersion')).primaryConnectionString]"
+      },
+      "DefaultSharedAccessPolicyPrimaryKey":{
+         "type":"string",
+         "value":"[listkeys(variables('defaultAuthRuleResourceId'), variables('serviceBusApiVersion')).primaryKey]"
+      },
+      "DefaultSqlConnectionString":{
```
Comment (gatneil) [inline 101-episerver-in-azure/azuredeploy.json]: Are we sure we want this as an output? Not sure if that's good security practice.

File: 101-episerver-in-azure/README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+[App Service Pricing](https://azure.microsoft.com/en-us/pricing/details/app-service/)
+
+[SQL Database Pricing](https://azure.microsoft.com/en-us/pricing/details/sql-database/)
+
+[Storage Service Pricing](https://azure.microsoft.com/en-us/pricing/details/storage/blobs/)
+
+[Service Bus Pricing](https://azure.microsoft.com/en-us/pricing/details/service-bus/)
+
+For more information about Running EPiServer in Azure, [Click Here](https://azure.microsoft.com/en-us/blog/announcing-episerver-cms-in-azure-marketplace-3/).
```
Comment (gatneil) [inline 101-episerver-in-azure/README.md]: It looks like this template only deploys the resources required to provision EpiServer, and it doesn't actually deploy episerver itself. Can we add some instructions or a link here on how to deploy episerver once this template has finished?

**Comment by gatneil** [comment]
Please address the above comments :)

File: 101-episerver-in-azure/README.md (hunk 7 above)
Comment (rajbalakrish) [inline 101-episerver-in-azure/README.md]: Indeed. I will add brief instructions on Deploying EpiserverCMS

File: 101-episerver-in-azure/azuredeploy.json (hunk 6 above)
Comment (rajbalakrish) [inline 101-episerver-in-azure/azuredeploy.json]: That's right. Included for local testing and can be removed

File: 101-episerver-in-azure/azuredeploy.json (hunk 5 above)
Comment (rajbalakrish) [inline 101-episerver-in-azure/azuredeploy.json]: Agreed. Will address in the next commit

**Comment by rajbalakrish** [comment]
@gatneil Thanks for your exhaustive review and hereby fixed the below items. Pl chcek if this is now Ok for a merge :) Thanks for all your patience 

1. Renamed folder name to episerver-cms-in-azure
2. Added Default Value to parameters to minimize user effort
3. Removed the Preview Apiversion and moved to stable Api versions
4. Removed Empty dependsOn List
5. Updated ReadMe with Links leading to step by step guidance on Deploying EpiserverCMS from VisualStudio

File: episerver-cms-in-azure/azuredeploy.parameters.json [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      "value": "1"
+    },
+    "StorageAccountName": {
+      "value": "GEN-UNIQUE"
+    },
+    "sqlserverName": {
+      "value": "GEN-UNIQUE"
+    },
+    "sqlserverAdminLogin": {
+      "value": "GEN-USER"
```
Comment (gatneil) [inline episerver-cms-in-azure/azuredeploy.parameters.json]: GEN-USER isn't an actual placeholder. For users, we usually do something generic like 'azureuser' or use GEN-UNIQUE

**Comment by gatneil** [comment]
One small change in the parameters file. Otherwise looks good :)

File: episerver-cms-in-azure/azuredeploy.parameters.json (hunk 8 above)
Comment (rajbalakrish) [inline episerver-cms-in-azure/azuredeploy.parameters.json]: Agreed :+1: Commit on the way !!!

**Comment by rajbalakrish** [comment]
@gatneil : Hereby Fixed :100:

**Comment by gatneil** [comment]
Looks good; merging; thanks for the contribution! :)

**Comment by rajbalakrish** [comment]
Many Thanks for your time and review @gatneil

**Pull request merged**